cores/arduino/Arduino.h

```diff
@@ -89,14 +89,40 @@ void yield(void);
 #undef abs
 #endif
 
-#define min(a,b) ((a)<(b)?(a):(b))
-#define max(a,b) ((a)>(b)?(a):(b))
-#define abs(x) ((x)>0?(x):-(x))
-#define constrain(amt,low,high) ((amt)<(low)?(low):((amt)>(high)?(high):(amt)))
-#define round(x)     ((x)>=0?(long)((x)+0.5):(long)((x)-0.5))
+#ifdef __GNUC__
+	#define max(a,b) \
+	({ typeof (a) _a = (a); \
+		typeof (b) _b = (b); \
+		_a > _b ? _a : _b; })
+	#define min(a,b) \
+	({ typeof (a) _a = (a); \
+		typeof (b) _b = (b); \
+		_a < _b ? _a : _b; })
+	#define abs(a) \
+	({ typeof (a) _a = (a); \
+		_a > 0 ? _a : _a * -1; })
+	#define constrain(amt, low, high) \
+	({ typeof (amt) _amt = (amt); \
+		typeof (low) _low = (low); \
+		typeof (high) _high = (high); \
+		_amt < _low ? _low : _amt > _high ? _high : _amt; })
+	#define round(x) \
+	({ typeof (x) _x = (x); \
+		((_x)>=0?(long)((_x)+0.5):(long)((_x)-0.5)) })
+	#define sq(x) \
+	({ typeof (x) _x = (x); \
+		_x * _x })
+#else
+	#define min(a,b) ((a)<(b)?(a):(b))
+	#define max(a,b) ((a)>(b)?(a):(b))
+	#define abs(x) ((x)>0?(x):-(x))
+	#define constrain(amt,low,high) ((amt)<(low)?(low):((amt)>(high)?(high):(amt)))
+	#define round(x)     ((x)>=0?(long)((x)+0.5):(long)((x)-0.5))
+	#define sq(x) ((x)*(x))
+#endif
+
 #define radians(deg) ((deg)*DEG_TO_RAD)
 #define degrees(rad) ((rad)*RAD_TO_DEG)
```

Comment:
These two lines (124, 125 for radians and degrees) expose the very same problem as do *all* macro calls. Why not define them as inline constexpr functions instead? Like so: 
'
`template<typename T>
inline constexpr 
T min(T a, T b) {
    return a > b ? b : a;
}`

Reply:
Because that doesn't work in C, and these macros are currently usable from C, so that would break compatibility.

Reply:
I was also concerned about binary size.